fili-core/src/test/groovy/com/yahoo/bard/webservice/web/handlers/WebServiceSelectorRequestHandlerSpec.groovy

```diff
@@ -85,6 +91,83 @@ class WebServiceSelectorRequestHandlerSpec extends Specification {
         5        |  null      | true            | 1
         null     |  1         | true            | 1
         5        |  1         | true            | 1
+    }
+
+    def "Test time function counts down based on total timing"() {
+        setup:
+
+        WebServiceSelectorRequestHandler handler = new WebServiceSelectorRequestHandler(
+                webService,
+                webServiceNext,
+                mapper,
+                TimeRemainingFunction.INSTANCE
+        )
+
+        RequestLog.startTiming(BardLoggingFilter.TOTAL_TIMER)
+        DataSource dataSource = Mock(DataSource)
+        GroupByQuery groupByQuery1 = new GroupByQuery(
+                dataSource,
+                DefaultTimeGrain.DAY,
+                Collections.emptyList(),
+                null,
+                null,
+                Collections.emptyList(),
+                Collections.emptyList(),
+                Collections.emptyList(),
+                null
+        )
+        Integer timeout = 20000
+        Integer timeSoFar = 0
+        QueryContext latestActualContext = null
+
+        and: "Stubbing"
+        DruidWebService onWebService
+        DataRequestHandler nextHandler = webServiceNext
+        ContainerRequestContext crc = Mock(ContainerRequestContext)
+        crc.getHeaders() >> headerMap
+        headerMap.add(DataApiRequestTypeIdentifier.CLIENT_HEADER_NAME, DataApiRequestTypeIdentifier.CLIENT_HEADER_VALUE)
+        headerMap.add("referer", "http://somewhere")
+
+        onWebService = webService
+        onWebService.getTimeout() >> timeout
+        onWebService.getServiceConfig() >> serviceConfig
+
+        nextHandler
+        rc = new RequestContext(crc, true)
+
+        when:
+        Thread.sleep(1000)
+        handler.handleRequest(rc, request, groupByQuery1, response)
+
+
+        then:
+        1 * nextHandler.handleRequest(rc, request, _, response) >> {
+            arguments ->
+                timeSoFar = (int) (RequestLog.fetchTiming(BardLoggingFilter.TOTAL_TIMER).getActiveDuration() / 1000000)
+                latestActualContext = ((GroupByQuery) arguments[2]).getContext()
+                return true
+        }
+        timeSoFar > 1000
+        timeSoFar < 19000
```

Comment:
may be do we also need to check the new timeout value here ? 
timeSoFar < 19000 && latestActualContext.getTimeout() < 19000